src/PasswordPolicy.php

```diff
@@ -0,0 +1,289 @@
+        $third_party_settings['minimumRequiredConstraints'] ?? 1,
+        $third_party_settings['constraints'] ?? []);
+    }
+    return NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function loadByRoles(array $roles): ?PasswordPolicyInterface {
+    $policies = \Drupal::service('password_enhancements.password_constraint')->loadMultiplePoliciesByRoles($roles);
+    return $policies ? reset($policies) : NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getExpireDays(): int {
+    return floor($this->getExpireSeconds() / 86400);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getExpireSeconds(): int {
+    return $this->expireSeconds;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getExpireWarnSeconds(): int {
+    return $this->expireWarnSeconds;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getExpireWarnDays(): int {
+    return floor($this->getExpireWarnSeconds() / 86400);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getExpiryWarningMessage(): ?string {
+    return $this->expiryWarningMessage;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getMinimumRequiredConstraints(): int {
+    return $this->minimumRequiredConstraints;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getPriority(): int {
+    return $this->role->getWeight();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getRole(): RoleInterface {
+    return $this->role;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setExpireSeconds(int $seconds): PasswordPolicyInterface {
+    if ($seconds < 0) {
+      throw new PolicyInvalidArgumentException('The expiry given in seconds must be a positive integer or zero.');
+    }
+    $this->expireSeconds = $seconds;
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setExpireWarnSeconds(int $seconds): PasswordPolicyInterface {
+    if ($seconds < 0) {
+      throw new PolicyInvalidArgumentException('The expiry warning in seconds must be a positive integer or zero.');
+    }
+    $this->expireWarnSeconds = $seconds;
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setExpiryWarningMessage(?string $message): PasswordPolicyInterface {
+    $this->expiryWarningMessage = $message;
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setMinimumRequiredConstraints(int $number): PasswordPolicyInterface {
+    if ($number < 0) {
+      throw new PolicyInvalidArgumentException('The minimum requirements cannot be less than zero.');
+    }
+    $this->minimumRequiredConstraints = $number;
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getConstraint($constraint): PasswordConstraintInterface {
+    return $this->getConstraints()->get($constraint);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getConstraints(): PasswordConstraintPluginCollection {
+    if (!$this->constraintsCollection) {
```

Comment:
This property could be null and it should not be null, this is the reason why you have a  workaround here. When a new object constructed this property must be initialized. BUT the real question is why this method returns a `PasswordConstraintPluginCollection` instead of an array of constraints? If we can get rid of the current solution we could also get rid of the `$this->getPasswordConstraintPluginManager()` that requires to load a service to a POJO object without DI.

Reply:
As that'd solve only one part of the problem, I don't want to get into that ATM.